lang/cs/Org.Apache.REEF.Network/Elastic/Task/Workflow.cs

```diff
@@ -0,0 +1,298 @@
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+using Org.Apache.REEF.Network.Elastic.Config;
+using Org.Apache.REEF.Network.Elastic.Failures;
+using Org.Apache.REEF.Network.Elastic.Operators;
+using Org.Apache.REEF.Network.Elastic.Operators.Physical;
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Utilities.Logging;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace Org.Apache.REEF.Network.Elastic.Task
+{
+    /// <summary>
+    /// Task-side representation of the the sequence of group communication operations to execute.
+    /// Exception rised during execution are managed by the framework and recovered through the user-defined
+    /// policies / mechanisms.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    public sealed class Workflow : IEnumerator<IElasticOperator>
+    {
+        private static readonly Logger LOGGER = Logger.GetLogger(typeof(Workflow));
+
+        private int _position = -1;
+        private bool _failed;
+        private bool _disposed;
+        private List<int> _iteratorsPosition;
+
+        private readonly object _lock;
+        private readonly IList<IElasticOperator> _operators;
+        private readonly CancellationSource _cancellationSource;
+        private readonly bool _isRescheduled;
+
+        /// <summary>
+        /// Injectable constructor.
+        /// </summary>
+        /// <param name="cancellationSource"></param>
+        [Inject]
+        private Workflow(
+            [Parameter(typeof(GroupCommunicationConfigurationOptions.IsRescheduled))] bool isRescheduled,
+            CancellationSource cancellationSource)
+        {
+            _operators = new List<IElasticOperator>();
+            _failed = false;
+            _disposed = false;
+            _lock = new object();
+            _iteratorsPosition = new List<int>();
+            _cancellationSource = cancellationSource;
+            _isRescheduled = isRescheduled;
+        }
+
+        /// <summary>
+        /// The current iteration value.
+        /// </summary>
+        public object Iteration
+        {
+            get
+            {
+                if (_iteratorsPosition.Count == 0)
+                {
+                    return 0;
```

Comment:
yea. I used object as a base type because in theory one could implement an iterator scanning anything (not just iterating over ints). Iterator should probably be typed and this piece should return the default value. Let's postpone this for when I will push the iterator operator.